src/config.c

```diff
@@ -801,68 +840,76 @@ void configSetCommand(client *c) {
+            errstr = "duplicate parameter";
+            err_arg_name = c->argv[2+i*2]->ptr;
+            invalid_args = 1;
+        } else {
+            dictAdd(set_configs, sdsnew(config->name), config);
+            dictAdd(new_values, sdsnew(config->name), c->argv[2+i*2+1]->ptr);
         }
-        set_configs[i] = config;
+        sdsfree(config_name);
         config_names[i] = config->name;
-        new_values[i] = c->argv[2+i*2+1]->ptr;
     }
     
     if (invalid_args) goto err;
 
     /* Backup old values before setting new ones */
-    for (i = 0; i < config_count; i++)
-        old_values[i] = set_configs[i]->interface.get(set_configs[i]);
+    dictIterator *di = dictGetIterator(set_configs);
+    dictEntry *de;
+    while ((de = dictNext(di)) != NULL) {
+        standardConfig *config = dictGetVal(de);
+        dictAdd(old_values, sdsnew(config->name), config->interface.get(config));
+    }
+    dictReleaseIterator(di);
 
     /* Set all new values (don't apply yet) */
     for (i = 0; i < config_count; i++) {
-        int res = performInterfaceSet(set_configs[i], new_values[i], &errstr);
+        sds config_name = sdsnew(config_names[i]);
+        standardConfig *config = dictGetVal(dictFind(set_configs, config_name));
+        sds new_value = dictGetVal(dictFind(new_values, config_name));
+        sdsfree(config_name);
+        int res = performInterfaceSet(config, new_value, &errstr);
         if (!res) {
-            restoreBackupConfig(set_configs, old_values, i+1, NULL, NULL);
-            err_arg_name = set_configs[i]->name;
+            restoreBackupConfig(set_configs, old_values, NULL, NULL);
+            err_arg_name = config_names[i];
             goto err;
         } else if (res == 1) {
             /* A new value was set, if this config has an apply function then store it for execution later */
-            if (set_configs[i]->flags & MODULE_CONFIG) {
-                addModuleConfigApply(module_configs_apply, set_configs[i]->privdata);
-            } else if (set_configs[i]->interface.apply) {
+            if (config->flags & MODULE_CONFIG) {
+                addModuleConfigApply(module_configs_apply, config->privdata);
+            } else if (config->interface.apply) {
                 /* Check if this apply function is already stored */
-                int exists = 0;
-                for (j = 0; apply_fns[j] != NULL && j <= i; j++) {
-                    if (apply_fns[j] == set_configs[i]->interface.apply) {
-                        exists = 1;
-                        break;
-                    }
-                }
-                /* Apply function not stored, store it */
-                if (!exists) {
-                    apply_fns[j] = set_configs[i]->interface.apply;
-                    config_map_fns[j] = i;
+                config_name = sdsnew(config->name);
+                if (!(dictFind(apply_fns, config_name))) {
```

Comment:
This seems wrong. The key to the `apply_fns` dict needs to be the function pointer not the config name. We need to store the the config_name as a the value for reporting later if the apply fails.

Reply:
Why would this be wrong?
Currently the code iterates over the dictionary and gets the function to apply by using the dictGeyVal function.

Reply:
The design is that you can't map a "config name" or `standardConfig` to an apply function and then call all the apply functions. We want to avoid calling the same apply function twice if there are two configs that use the same apply function. So we need to collect all the **unique** apply functions and call them later after doing the set (or when rolling back a failed set to old values). This means that the unique value in the `apply_fns` dict needs to be the function pointer itself, not the config name.

Reply:
Yes that makes sense, thank you. I have updated the code now so it takes the function pointer as the key.